artifacts.go

```diff
@@ -301,205 +294,486 @@ func validateArtifact(c *cli.Context) error {
+	art, err := os.Open(artifact)
+	if err != nil {
+		return err
+	}
+	defer art.Close()
+
+	tmp, err := ioutil.TempFile("", "mender-artifact")
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tmp.Name())
+
+	var privateKey []byte
+	if key != "" {
+		privateKey, err = getKey(key)
+		if err != nil {
+
+			return cli.NewExitError(fmt.Sprintf("Can not use signing key provided: %s", err.Error()), 1)
+		}
+	}
+
+	if _, err = repack(art, tmp, privateKey, newName, rootfs); err != nil {
 		return err
 	}
 
-	// read the rest of the artifact
-	for {
-		raw, err := r.ReadRaw()
-		if err != nil && errors.Cause(err) == io.EOF {
-			break
-		} else if err != nil {
+	return os.Rename(tmp.Name(), artifact)
+}
+
+func modifyExisting(c *cli.Context, image string) error {
+	if c.String("name") != "" {
+		if err := modifyName(c.String("name"), image); err != nil {
 			return err
 		}
-		if err = w.WriteRaw(raw); err != nil {
+	}
+
+	if c.String("server-uri") != "" {
+		if err := modifyMenderConfVar("ServerURL",
+			c.String("server-uri"), image); err != nil {
+			return err
+		}
+	}
+
+	if c.String("server-cert") != "" {
+		if err := modifyServerCert(c.String("server-cert"), image); err != nil {
+			return err
+		}
+	}
+
+	if c.String("verification-key") != "" {
+		if err := modifyVerificationKey(c.String("verification-key"), image); err != nil {
 			return err
 		}
 	}
+
+	if c.String("tenant-token") != "" {
+		if err := modifyMenderConfVar("TenantToken",
+			c.String("tenant-token"), image); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }
 
-func writeTemp(aName string, key []byte, force bool) (string, error) {
+type partition struct {
+	offset string
+	size   string
+	path   string
+}
 
-	f, err := os.Open(aName)
+func processSdimg(image string) ([]partition, error) {
+	out, err := exec.Command("parted", image, "unit s", "print").Output()
 	if err != nil {
-		return "", errors.Wrapf(err, "Can not open: %s", aName)
+		return nil, errors.Wrap(err, "can not execute `parted` command or image is broken; "+
+			"make sure parted is available in your system and is in the $PATH")
 	}
-	defer f.Close()
 
-	// initialize raw reader and writer
-	aReader := areader.NewReader(f)
-	ver, data, err := aReader.ReadRawVersion()
-	if err != nil {
-		return "", err
-	}
+	partitions := make([]partition, 0)
 
-	// we are supporting only v2 signing for now
-	switch ver {
-	case 1:
-		return "", errors.New("Can not sign v1 artifact")
-	case 2:
-		tFile, err := ioutil.TempFile("", "mender-artifact")
-		if err != nil {
-			return "", errors.Wrap(err,
-				"Can not create temporary file for storing artifact")
+	reg := regexp.MustCompile(`(?m)^[[:blank:]][0-9]+[[:blank:]]+([0-9]+)s[[:blank:]]+[0-9]+s[[:blank:]]+([0-9]+)s`)
+	partitionMatch := reg.FindAllStringSubmatch(string(out), -1)
+
+	// IMPORTANT: we are assuming standard Mender formating here:
+	// only 2nd and 3rd partitions are rootfs we are going to modify
```

Comment:
Not sure how much work it is, but there is a simple test/assertion you can do: Just verify that `/etc/mender/artifact_info` exists on the partition.

Reply:
It is done elsewhere, where we are actually modifying something. Here we would need to dd the partitions to some temporary place and check if the file exists. Might be quite heavy especially because it is verified later on anyway.